src/tools/database/sql.ts

```diff
@@ -0,0 +1,116 @@
+ * limitations under the License.
+ */
+
+import {
+  Tool,
+  ToolInput,
+  ToolError,
+  BaseToolOptions,
+  BaseToolRunOptions,
+  JSONToolOutput,
+} from "@/tools/base.js";
+import { z } from "zod";
+import { Sequelize } from "sequelize";
+import { getMetadata } from "@/tools/database/metadata.js";
+import { connectSql } from "@/tools/database/connection.js";
+import { Cache } from "@/cache/decoratorCache.js";
+
+type Provider = "mysql" | "mariadb" | "postgres" | "mssql" | "db2" | "sqlite" | "oracle";
+
+type ToolOptions = { provider: Provider } & BaseToolOptions;
+type ToolRunOptions = BaseToolRunOptions;
+
+export class SQLTool extends Tool<JSONToolOutput<any>, ToolOptions, ToolRunOptions> {
+  name = "SQLTool";
+
+  description =
+    "Converts natural language to SQL query and executes it using the provided tables schema.";
+
+  inputSchema() {
+    return z.object({
+      query: z.string({ description: "The SQL query to be executed." }).min(1),
+    });
+  }
+
+  static {
+    this.register();
+  }
+
+  private _sequelize: Sequelize | null = null;
+
+  @Cache()
+  protected async connection(): Promise<Sequelize> {
+    this._sequelize = await connectSql();
+    return this._sequelize;
+  }
+
+  protected async _run(
+    { query }: ToolInput<this>,
+    _options?: ToolRunOptions,
+  ): Promise<JSONToolOutput<any>> {
+    if (!this.isReadOnlyQuery(query)) {
+      return new JSONToolOutput({
+        success: false,
+        error: "Invalid query. Only SELECT queries are allowed.",
+      });
+    }
+
+    let metadata = "";
+    const provider = this.options.provider;
+    const sequelize = await this.connection();
+
+    try {
+      metadata = await getMetadata(provider, sequelize);
+
+      const [results] = await sequelize.query(query);
+
+      if (Array.isArray(results) && results.length > 0) {
+        return new JSONToolOutput({ success: true, results });
+      } else {
+        return new JSONToolOutput({
+          success: false,
+          message: `No rows selected`,
+        });
+      }
+    } catch (error) {
+      const errorMessage = `Based on this database schema structure: ${metadata}, 
+      generate a correct query that retrieves data using the appropriate ${provider} dialect. 
+      The original request was: ${query}, and the error was: ${error.message}.`;
+
+      throw new ToolError(errorMessage);
+    }
+  }
+
+  private isReadOnlyQuery(query: string): boolean {
+    const normalizedQuery = query.trim().toUpperCase();
+    return (
+      normalizedQuery.startsWith("SELECT") ||
+      normalizedQuery.startsWith("SHOW") ||
+      normalizedQuery.startsWith("DESC")
+    );
+  }
+
+  public async destroy(): Promise<void> {
+    if (this._sequelize) {
+      try {
+        await this._sequelize.close();
+        this._sequelize = null;
+      } catch (error: any) {
+        throw new ToolError(`Failed to close the database connection: ${error.message}`);
+      }
+    }
+  }
```

Comment:
Rebase against the `main` branch and do the following.

```typescript
public async destroy() {
   // @ts-expect-error protected property
   const cache = Cache.getInstance(this, "connection")
   const entry = cache.get()

   if (entry) {
      cache.clear()
      
      try {
        await entry.data.close();
      } catch (error) {
        throw new ToolError(`Failed to close the database connection`, [error]);
      }
   }
}
```

Reply:
this doesn't close the connection when process.exit(0) follows

Reply:
There might be a typo in the code. But it does the same thing as your previous code.